ls1intum/Artemis
# [Feature] Support for PMD Copy/Paste Detection

### Checklist

### Motivation and Context


Clones are a threat to the maintainability of software systems. Pieces of code which are used often should be encapsulated and made available to other components by providing well documented interfaces (Don't repeat yourself, Single responsibility principle). With this PR, code clones can be automatically detected and punished by configuring the grading for programming exercises.

### Description


- Add `PMD CPD` tool, which automatically extends the build plan if SCA is enabled
- Add Copy Paste detection category
- Add default configuration for PMD CPD. Only duplication with a minimum size of 70 tokens are reported.

### Steps for Testing


**Can only be tested on TS2 at the moment, as this PR relies on a new version of the Notification Plugin, which is already deployed on Jenkins**
1. Create a programming exercise with SCA enabled
2. Participate
3. Submit code with contains duplications (must be at least 70 tokens long, this can be changed in the pom, the BubbleSort algorithm should suffice)
--> Check that the code duplication is detected

### Test Coverage

Nothing changed

### Screenshots


![image](https://user-images.githubusercontent.com/16407766/97510877-fbc0e680-1985-11eb-862f-49c2ab9887a4.png)
